### PR TITLE
RDM-10876

### PR DESCRIPTION
**Pull request by tomxelliott**
Implement cache for caseRoles and challengeQuestions calls to Definition Store

**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDM-10876

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/functionalTest/resources/features/F-205 - Validate NoC Questions/F-205.feature [hunk 1]
```diff
@@ -16,6 +16,7 @@ Feature: F-205 Validate Notice of Change Answers
       And a case [created by Richard - C1 - on behalf of Mario as the applicant which is auto-assigned to Richard's organisation] created as in [204_Case_Creation],
       And a successful call [to get the update org policy event trigger] as in [204_Update_Org_Policies_Token_Creation],
       And a successful call [to update the applicant org policy] as in [204_Update_Org_Policies],
+      And a wait time of [15] seconds [to allow for the cache to refresh],
```
Comment (MSancaktutar) [inline src/functionalTest/resources/features/F-205 - Validate NoC Questions/F-205.feature:19]: You could describe this wait time in more none techie way...

like wait time for latest xyz definition data to take effect for abc function

Comment (tomxelliott) [reply src/functionalTest/resources/features/F-205 - Validate NoC Questions/F-205.feature:19]: Yeah good idea, will take a look at it 👍

File: src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDefinitionStoreRepository.java [hunk 2]
```diff
@@ -8,11 +8,14 @@
...
 import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.Maps.newHashMap;
 
 @Repository
 public class DefaultDefinitionStoreRepository implements DefinitionStoreRepository {
 
-
+    public final Map<String, Integer> versions = newHashMap();
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDefinitionStoreRepository.java:18]: Hashmap is not thread safe. Use concurrenthashmap please, if you need a shared map instance across requests.

File: src/functionalTest/resources/features/F-204 - Get NoC Questions/F-204.feature [hunk 3]
```diff
@@ -144,6 +144,7 @@ Feature: F-204 Get Notice of Change Questions
       And a successful call [to update the applicant org policy] as in [204_Update_Org_Policies_With_Missing_Policy],
       And [The challenge questions have answers that resolve to three distinct case roles (1,2,3)] in the context of the scenario
       And [The configuration of the case creation event only establishes organisation policies containing case roles 1 and 2] in the context of the scenario
+      And a wait time of [15] seconds [to allow for the cache to refresh],
```
Comment (smathangi) [inline src/functionalTest/resources/features/F-204 - Get NoC Questions/F-204.feature:147]: why wait..? i thought with caseType version you never get stale data.  In any case 15s a lot which increases build (as we have so many NoC scenarios) ?

Comment (tomxelliott) [reply src/functionalTest/resources/features/F-204 - Get NoC Questions/F-204.feature:147]: The BEFTA tests are having issues with the cached data - 19 failures in total. Only these 19 tests have the wait period.

File: src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDefinitionStoreRepository.java [hunk 4]
```diff
@@ -21,13 +24,23 @@ public DefaultDefinitionStoreRepository(DefinitionStoreApiClient definitionStore
     }
 
     @Override
+    @Cacheable(value = "challengeQuestions",
+        condition = "#root.target.getLatestVersion(#caseTypeId) == #root.target.versions[#caseTypeId]")
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDefinitionStoreRepository.java:28]: I guess cache condition/unless are for evaluating whether result should cache or not 🤔  
May be worth to add some cache IT tests for this class?

**Comment by MSancaktutar** [comment]
@tomxelliott ,
We need to make sure the FTs are executing in 10 parallel threads, as otherwise those wait times will dramatically increase the total FT execution times.
If 10 threads run, overall extension in total time for 19 wait lines will be around 30 seconds.

File: src/main/java/uk/gov/hmcts/reform/managecase/repository/CaseTypeDefinitionVersion.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.io.Serializable;
+
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+public class CaseTypeDefinitionVersion implements Serializable {
+
+    private static final long serialVersionUID = 3792842101045258030L;
+
+    private Integer version;
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/managecase/repository/CaseTypeDefinitionVersion.java:11]: May be lombok annotations?

File: src/integrationTest/java/uk/gov/hmcts/reform/managecase/repository/DefinitionStoreRepositoryIT.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+@RunWith(SpringRunner.class)
+@ContextConfiguration
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
```
Comment (smathangi) [inline src/integrationTest/java/uk/gov/hmcts/reform/managecase/repository/DefinitionStoreRepositoryIT.java:31]: do we really need this ? Try resetting caches/ cacheManager in the teardown()

File: src/integrationTest/java/uk/gov/hmcts/reform/managecase/repository/DefinitionStoreRepositoryIT.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            initMocks(this);
+            FieldType fieldType = FieldType.builder()
+                .regularExpression("regular expression")
+                .max(null)
+                .min(null)
+                .id("Number")
+                .type("Number")
+                .build();
+            ChallengeQuestion challengeQuestion = ChallengeQuestion.builder()
+                .caseTypeId(CASE_TYPE_ID)
```
Comment (smathangi) [inline src/integrationTest/java/uk/gov/hmcts/reform/managecase/repository/DefinitionStoreRepositoryIT.java:57]: move all these into @Before setup() method

File: src/main/java/uk/gov/hmcts/reform/managecase/repository/CaseTypeDefinitionVersion.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import java.io.Serializable;
+
+import lombok.Getter;
+import lombok.Setter;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+@Getter
+@Setter
+public class CaseTypeDefinitionVersion implements Serializable {
+
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/managecase/repository/CaseTypeDefinitionVersion.java:12]: May be simple @Data annotation?  That would give all other required ones including toString()

**Review by smathangi — APPROVED** [review]
LGTM. (few minor comments)

**Pull request closed without merging**